### DISCUSSION TITLE
Accessing CONNACK properties

### DISCUSSION THREAD

**Discussion by airtower-luna** [empicano/aiomqtt#338] [Q&A]
I'm looking to access MQTT v5 properties sent by the server in CONNACK, primarily [Response Information](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Response_Information), but access to others may be useful, too. The only way I found is hooking into `aiomqtt.Client._on_connect()`, e.g.:

```py
class PropClient(aiomqtt.Client):
    def _on_connect(
            self,
            client: mqtt.Client,
            userdata: Any,
            flags: mqtt.ConnectFlags,
            reason_code: ReasonCode,
            properties: Properties | None = None) -> None:
        super()._on_connect(
            client, userdata, flags, reason_code, properties)
        self.connack_properties = properties
```

This is obviously undesirable because it's using (and modifying) internal API. Am I missing something, or is this currently missing from the API? This seems like a key piece for MQTT v5 request/response.

**Comment by empicano** [comment]
Hi there! Yes, we're currently ignoring the CONNACK properties completely. I'd be very happy to see request/response fully functional in aiomqtt 🙂

For PUBLISH messages we receive from the broker, we currently expose _all_ properties via `aiomqtt.Message.properties`, which simply returns paho's `Properties` type. I don't think that's the best way to do it. Something like the subscription identifier property is probably not interesting to the user and should thus not be unnecessarily exposed.

I'd like to only expose CONNACK properties that are useful to the user. What CONNACK properties apart from response information do you think might be useful to access?

---

Some unsorted thoughts:

- CONNACK properties will only be available inside the context manager, defining them as standard properties of `Client` might make the typing cumbersome, because it's going to be a union with None
- CONNACK properties might change on reconnection (?)
- Can we make the interface to send requests (and receive responses -- see also #304) or the interface to send responses more straightforward than it is now without loosing functionality?

I'd be super happy to hear what you think!

**Reply by airtower-luna** [reply]
> I'd like to only expose CONNACK properties that are useful to the user. What CONNACK properties apart from response information do you think might be useful to access?

*So far* Response Information is the only one I need, but I'm still relatively new to MQTT, so I don't know what else I might want to use in the future. With `client.publish()` and `client.messages` the existing properties interface allowed me to build functional request/response code (except for deriving the Response Topic from Response Information provided by the broker), so while I'd like to see a better API I think having that low level access if needed is valuable. :wink:

That applies even more for tests, where I might want to see low level data simply to make sure it's correct, or mess with it to test error cases.

> CONNACK properties will only be available inside the context manager, defining them as standard properties of `Client` might make the typing cumbersome, because it's going to be a union with None

A broker doesn't *have* to provide Response Information, so that case must be handled either way. If there are enough properties that are mandatory (I haven't checked yet), maybe having a subclass of `Client` for v5 connections would make sense?

> Can we make the interface to send requests (and receive responses -- see also https://github.com/empicano/aiomqtt/pull/304) or the interface to send responses more straightforward than it is now without loosing functionality?

*Ideally* I'd like to have a `Client.request()` method, which returns an Awaitable that gives me the response when it resolves. Considerations:

* Correlation Data could be set automatically with random bytes (at least by default, not sure if there might be a need to override it).
* Response Topic is more complicated. If the broker has provided Response Information, it would be possible to automatically add a random suffix to the namespace. If it doesn't the user would at least have to provide the namespace to use (maybe optionally set it on the connected client?). Not sure if someone might need to force a specific response topic under some circumstances, but especially for access control it might make sense to extend the base namespace for responses.
* The routing feature would probably make implementing a `Client.request()` kind of interface much simpler (add a route right before sending the request, remove it after receiving the response). :slightly_smiling_face:
* On the side that's sending a response, it'd be nice for the `Message` object to offer a `respond` or `reply` method similar to `publish` that automatically sets topic and correlation data to match the request. Maybe subclassing would be an option here too, so code could detect requests with `isinstance()` and then handle them accordingly?

While talking about MQTTv5 properties, something that should be pretty simple to solve is using Content Type. Currently I have to manually create a Properties object, set PayloadFormatIndicator and ContentType, and the pass the object to publish. Could be a pretty short PR to add a `content_type` parameter to publish, and expose it on the `Message` object. The main question would be what to do if the `properties` parameter to `publish` is set and contains a content type. ValueError?

**Reply by empicano** [reply]
Really really good points, thanks for that! 👍

> while I'd like to see a better API I think having that low level access if needed is valuable. That applies even more for tests, where I might want to see low level data simply to make sure it's correct, or mess with it to test error cases.

That makes sense. It's probably more that I don't like the current `Properties` interface that we currently take directly from paho, than that we expose all of the properties. I do think that we should nonetheless aim to minimize the _need_ to access these properties, because that usually means that the interface is lacking 🙂

> _Ideally_ I'd like to have a Client.request() method, which returns an Awaitable that gives me the response when it resolves.

Request/response for MQTT is a little different than for HTTP., because we basically "request" into the void: There's no guarantee that we even get a response and, on the other hand, we also might get more than one response. We could think about something like this:

```python
async with asyncio.timeout(10):  # How long do we want to wait for responses?
    async for response in client.request(...):
        print(response.payload)
```

However, that starts to introduce a second interface of handling messages. For the sake of keeping things simple, I'd hope to integrate request/response into what #304 proposes, if that's possible. In the past, we had a similar interface for "normal" messages and thus had multiple queues, which brought many annoying problems with them (e.g. controlling overall queue size). Another one is that #304 makes concurrent message handling with multiple workers very straightforward. What do you think?

> The routing feature would probably make implementing a Client.request() kind of interface much simpler (add a route right before sending the request, remove it after receiving the response)

I always separate subscriptions and routing in my head. #304 keeps subscriptions "dynamic" (we can subscribe/unsubscribe whenever we want) and implements routing "static" (how messages are handled is fixed and defined when creating the `Client` object). Does that clash with your use case somehow?

We do need some idea of how users can match incoming responses with their respective requests, though.

> On the side that's sending a response, it'd be nice for the Message object to offer a respond or reply method similar to publish that automatically sets topic and correlation data to match the request. Maybe subclassing would be an option here too, so code could detect requests with isinstance() and then handle them accordingly?

That sounds like a good idea and doesn't sound like we'd be loosing any functionality through abstraction 👍 Another idea could be to implement an optional parameter to the existing `publish` to pass a message we want to reply to. Similar thought as above with trying to avoid splitting the interface in two.

**Reply by airtower-luna** [reply]
> I always separate subscriptions and routing in my head. https://github.com/empicano/aiomqtt/pull/304 keeps subscriptions "dynamic" (we can subscribe/unsubscribe whenever we want) and implements routing "static" (how messages are handled is fixed and defined when creating the Client object). Does that clash with your use case somehow?

The problem is that the response topic isn't necessarily static. The spec doesn't really place any requirements, but mentions config, CONNACK Response Information property (which is why I want access to that, but not every broker supports/provides it), and randomization (for the last part of a Response Topic) as possible sources. See: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901254

I suppose whether it clashes depends on what you want to do with the routing here, registering a route for a base topic on CONNACK (using either Response Information or config) to catch any responses for further processing might work. But then any function calls to send requests should ensure the Response Topic is actually in the correct namespace.

> We do need some idea of how users can match incoming responses with their respective requests, though.

If you want to keep it similar to existing code, an analogue to `message.topic.matches()` that matches Correlation Data might work. But that would require the user to keep track of the Correlation Data or some request object holding it. I'm not convinced that's simpler than a separate queue (like the `client.request()` idea), at least on the user side, inside the Client object may be a different story. :sweat_smile: 

> That sounds like a good idea and doesn't sound like we'd be loosing any functionality through abstraction 👍 Another idea could be to implement an optional parameter to the existing publish to pass a message we want to reply to. Similar thought as above with trying to avoid splitting the interface in two.

Would work, but seems rather counterintuitive to me, especially that the topic (which is otherwise mandatory) would suddenly be filled from the "reply to this" object (plus setting Correlation Data). I guess it'd be a bit less confusing with passing the message to reply to *instead of* the topic, but having the same parameter take on different meanings still seems confusing.

**Reply by empicano** [reply]
> The problem is that the response topic isn't necessarily static.

Huh, you're right 👍 I'm not sure that letting aiomqtt handle Response Information is the right way, given that the specification doesn't actually specify the contents. The problem with `message.topic.matches()` is that it's not a very efficient way to filter messages, that's one of the things I want to address with #304. And I'd really like to avoid going back to having multiple queues, they're just loaded with problems (e.g. how to limit combined queue size).

Piggybacking again on that PR (you can see that I'm trying to press request/response into the publish/subscribe interface, also because request/response is simply used by the minority of MQTT users, and I'd like to keep the interface as simple/non-confusing as possible), we could implement optional handlers on a request-basis:

```python
async def handle_foo(message, client):
    print("foo:", message.payload)


async def handle_bar(message, client):
    print("bar:", message.payload)


async with aiomqtt.Client(
    "test.mosquitto.org",
    handlers={"humidity/+/inside/#": handle_foo}  # That's the standard way to define handlers
) as client:
    await client.subscribe("humidity/#")
    await client.subscribe("response")
    # But we can also optionally set response topics and response handlers on an outgoing message
    await client.publish("request", response_topic="response", response_handler=handle_bar)
    async for message in client.messages:
        await client.route(message)
```

Because adding handlers dynamically is actually not a problem. Inside the client, we can then keep a mapping of correlation data to handlers. What do you think about that? Do you see that lacking somewhere? 🤔

> Would work, but seems rather counterintuitive to me, especially that the topic (which is otherwise mandatory) would suddenly be filled from the "reply to this" object (plus setting Correlation Data).

Makes sense 👍 If we go with the above, I think we could implement both request and response by adding some optional arguments to `Client.publish()`, but I think you convinced me that it's a good idea to have proper `request()` and `respond()` convenience functions. I think that I'd like to have the `respond()` function as member of the client, not as member of the message, though, somehow it doesn't seem clean to me to pass the client around as part of each message.

So, something like:

```python
async def handle_foo(message, client):
    if isinstance(message, Request):
        client.respond(message, "example")


async def handle_bar(message, client):
    print("bar:", message.payload)


async with aiomqtt.Client(
    "test.mosquitto.org",
    handlers={"humidity/+/inside/#": handle_foo}  # That's the standard way to define handlers
) as client:
    await client.subscribe("humidity/#")
    await client.subscribe("response")
    # But we can also optionally set response topics and response handlers on an outgoing message
    await client.request("request", response_topic="response", response_handler=handle_bar)
    async for message in client.messages:
        await client.route(message)
```

Thanks for discussing this so deeply with me, you're raising a lot of great points! 🙏

**Reply by airtower-luna** [reply]
> Because adding handlers dynamically is actually not a problem. Inside the client, we can then keep a mapping of correlation data to handlers. What do you think about that? Do you see that lacking somewhere? 🤔

The main issue I see there is: How long should response handlers stay enabled? If they're never removed the mapping will grow indefinitely as requests are sent. On the other hand as you pointed out before we might want to handle multiple responses per request, so removing a handler after the first response wouldn't be correct either, and if it was, we'd still need to handle the possibility that we never get any response. That's what the appeal of iterating using `async for` like for `client.messages` would be to me: It'd be easy to wrap that in a timeout, or `break` on some condition (though we'd still need to trigger cleanup on leaving the loop).

I guess we'd need some kind of handle that can be used to remove the request handler, possibly to be used as async context manager that does that on `__aexit__`? Or maybe for each request create a task that'll remove the response handler after a timeout, or earlier if cancelled (that'd require returning the task object from the request function)? Timeout seems suitable here because it allows making requests without worrying about explicit cleanup, as long as the timeout is short enough that the handler mapping won't grow to an unreasonable size (whatever that means in a given context), and I assume most people would want to use one for waiting on responses anyway.

Another thing is that people might want to pass additional data to handlers and use the results elsewhere, but that could be solved with closures, `functools.partial()`, or by passing a method of an object that holds the necessary state, so it doesn't necessarily have to be part of the API.

> Makes sense 👍 If we go with the above, I think we could implement both request and response by adding some optional arguments to Client.publish(), but I think you convinced me that it's a good idea to have proper request() and respond() convenience functions. I think that I'd like to have the respond() function as member of the client, not as member of the message, though, somehow it doesn't seem clean to me to pass the client around as part of each message.

Sounds good to me! :slightly_smiling_face:

**Reply by empicano** [reply]
> The main issue I see there is: How long should response handlers stay enabled?

That's a good point, and you're right that it would be easier to do that with an `async for` loop. I don't see a way how we can could have that _and_ not use additional queues, and I think the single queue is really important 🙁

I'm thinking that the reasons we would want to remove a response handler should be relatively few. I can think of:

- Timeout (e.g. stop handling responses after 10s)
- Limited number of messages (e.g. stop after the first message)
- Certain type of message (e.g. stop after we get a `stop=True` message)

We could implement the first two as optional arguments to `Client.request()` e.g. `Client.request("example", response_topic="response/topic", response_handler=handle_bar, timeout=10)`, for the third one, we could make the message router of the client directly accessible, so that the user can call `Client.router.remove("response/topic")` from inside the message handler.

I feel like that somehow leads us down a bad rabbit hole, though. Maybe the optional handlers on a request-basis that I proposed are already one step too far, and we should keep it simpler. How about:

- Add `Client.Router.insert()` and `Client.Router.remove()` functions to add/remove message handlers together with #304
- Add `Client.request()` and `Client.respond/reply()` or `Request.respond/reply()` convenience functions
- Expose CONNACK properties as properties of the `Client`

That's kind of chickening out from tracking correlation data and automating handler expiry and letting the user handle that instead, I admit that I don't have a good solution at the moment that still let's us keep the single queue. But it gives a few tools to make request/reponse easier, and shouldn't abstract so much that we're loosing functionality. The last point also addresses your original problem with regarding Response Information.

> Another thing is that people might want to pass additional data to handlers and use the results elsewhere

You already mention a few options, can you give an example of how your preferred solution would look like? You already proposed to attach the client to the message object instead of passing it as the second argument to the handlers. We could take this one step further and do something like [Starlette's lifespan state](https://www.starlette.io/lifespan/#lifespan-state) to attach a user-defined "state" object to each message.

**Reply by airtower-luna** [reply]
> I feel like that somehow leads us down a bad rabbit hole, though. Maybe the optional handlers on a request-basis that I proposed are already one step too far, and we should keep it simpler. How about:
> 
> * Add `Client.Router.insert()` and `Client.Router.remove()` functions to add/remove message handlers together with #304
> * Add `Client.request()` and `Client.respond/reply()` or `Request.respond/reply()` convenience functions
> * Expose CONNACK properties as properties of the `Client`

That seems reasonable to me. Together with V5 content type support (which should be fairly trivial) that'd simplify my code a lot already.

The one thing that seems odd to me is that I'd have to subscribe and set up routing for the response topic separately on the client that sends the request, unless I want to use the exact same handler for each request.

> You already mention a few options, can you give an example of how your preferred solution would look like? You already proposed to attach the client to the message object instead of passing it as the second argument to the handlers. We could take this one step further and do something like [Starlette's lifespan state](https://www.starlette.io/lifespan/#lifespan-state) to attach a user-defined "state" object to each message.

What I want is the same thing I can do with any thread, pass arguments to the handler function. Let's say I want to *request* some status report, and then process all the results, the code might end up roughly like this:

```python3
async def handle_report(message, client, corr: bytes, q: asyncio.Queue):
    if not hasattr(message.properties, 'CorrelationData') \
            or message.properties.CorrelationData != corr:
        # not for us
        return
    data = json.loads(message.payload)
    await q.put(data)


async def process_reports(client, q, r_topic):
    try:
        async with asyncio.timeout(10):
            data = await q.get()
            # do stuff with the data
    except TimeoutError:
        # okay, done.
        pass
    await client.unsubscribe(r_topic)
    client.router.remove(r_topic)
    return ...


async def keep_routing(client):
    async for message in client.messages:
        await client.route(message)


async with aiomqtt.Client('test.mosquitto.org') as client:
    q = asyncio.Queue()
    corr = secrets.token_bytes(8)
    # might use Response Information from CONNACK
    r_topic = f'response/status/{secrets.token_hex(8)}'
    client.router.insert(r_topic, handle_message, args=(corr, q))
    await client.subscribe(r_topic)
    await client.request(
        'device/send-status', response_topic=r_topic, correlation_data=corr)
    task1 = asyncio.create_task(process_reports(client, q, r_topic))
    task2 = asyncio.create_task(keep_routing(client))
    stuff = await task1
    task2.cancel()
```

Using `kwargs` might make more sense, but point is, I need to be able to pass data around, including to check the correlation data. Maybe I could limit the subscription to `response/status/+` and leave it active, but I think you can see my point here.

**Reply by empicano** [reply]
> The one thing that seems odd to me is that I'd have to subscribe and set up routing for the response topic separately on the client that sends the request, unless I want to use the exact same handler for each request.

We basically have three options for routing, that differ in how we handle overlapping subscriptions:

- We can route a message to the most specific handler that matches
- We can route it to all handlers that match
- We could [assign each subscription a specific handler](https://github.com/empicano/aiomqtt/pull/304#issuecomment-2496250860). Most brokers support subscription IDs, which would let us then match incoming messages to the exact subscription(s) -- and thus handler(s) -- that caused it

For reference, the MQTT spec states:

> When Clients make subscriptions with Topic Filters that include wildcards, it is possible for a Client’s subscriptions to overlap so that a published message might match multiple filters. In this case the Server MUST deliver the message to the Client respecting the maximum QoS of all the matching subscriptions [MQTT-3.3.4-2]. In addition, the Server MAY deliver further copies of the message, one for each additional matching subscription and respecting the subscription’s QoS in each case.

Leveraging subscription IDs would probably be the cleanest way to go. What I like about the first option is the simple interface:

```py
async for message in client.messages:
    await client.route(message)
```

Otherwise, we would have to deal with potentially multiple handlers:

```py
async for message in client.messages:
    for handler in client.route(message):
        await handler(message)
```

Although, maybe we might be able to duplicate incoming messages that have multiple subscription IDs assigned? That would standardize how aiomqtt works, even with different brokers.

We're really touching all details here, I like it 😄

**Reply by airtower-luna** [reply]
> * We can route a message to the most specific handler that matches

That seems more intuitive than "all matching handlers" to me.

> * We could [assign each subscription a specific handler](https://github.com/empicano/aiomqtt/pull/304#issuecomment-2496250860). Most brokers support subscription IDs, which would let us then match incoming messages to the exact subscription(s) -- and thus handler(s) -- that caused it

I agree, that seems like the most elegant solution, because it avoids having to set up subscription and routing separately when making a request. It'll require a slight change to `Client.unsubscribe()` to allow cancelling subscriptions by ID (instead of topic). I'm thinking of making a request, and canceling the response subscription (and thus handler) when done processing responses by whatever logic fits in the given context. This would solve the issue of being able to parametrize response handlers.

I'm not sure how it'd integrate with your routing concept, though. Would a subscription-specific handler override the default routing?

> Although, maybe we might be able to duplicate incoming messages that have multiple subscription IDs assigned? That would standardize how aiomqtt works, even with different brokers.

That seems useful, otherwise code might simply break depending on the broker it's used with, e.g. if it expects to receive messages for all matching handlers, but the broker sends the message only once. Either way I think it'd be better to handle that internally in the routing code, and not require user code to add boilerplate for calling handlers.
